---
sort: 1
---

# Welcome to CGMegaâ€™s documentation

CGMega is a graph attention-based deep learning framework for cancer gene module dissection. CGMega leverages a combination of multi-omics data cross genome, epigenome, proteome and especially three-dimension (3D) genome levels.

Source code is availabel at [CGMega GitHub repository(no link)](https://github.com) . For more information, please read our paper: [CGMega: dissecting cancer gene module with explainable graph attention network(no link)](https://www.nature.com/ncomms/).

>
> This website is still under the beta test phase. Thanks for your interest.
>

[TOC]

## [Index](https://sunyolo.github.io/CGMega.github.io/)
- [Installation](https://sunyolo.github.io/CGMega.github.io/test_long/folder1-Installation/)
	- [System Requirements](https://sunyolo.github.io/CGMega.github.io/test_long/folder1-Installation/)
	- [Software Requirements](https://sunyolo.github.io/CGMega.github.io/test_long/folder1-Installation/)
	- [Installation](https://sunyolo.github.io/CGMega.github.io/test_long/folder1-Installation/)
- [Tutorial](https://sunyolo.github.io/CGMega.github.io/test_long/folder2-Tutorial/)
	- [How to prepare input data](#1.how to prepare input data)
	- [Model training](https://sunyolo.github.io/CGMega.github.io/test_long/folder2-Tutorial/)
	- [Interpretation](#3.Model training)
